### PR TITLE
#402: Hash calculation improvements

### PR DESCRIPTION
**Pull request by mvanaken**
See also issue #402 , which explains the problem and possible solutions.

This PR implements hash caching for all immutable objects, which basically means for all objects within metal, so many files are adjusted. In short:
* [ImmutableObject](https://github.com/parsingdata/metal/pull/403/files#diff-81f3e3696155918279f269adae29f82bb381dcfbe181b589286a6d269a4d86e5) caches the hash so the hash is only calculated once.
* All immutable objects within metal extends this object
* [A test](https://github.com/parsingdata/metal/pull/403/files#diff-2de103cc58b5476f76502f919b02ba47b83ff2528554d554d9022e8ff6319fd3) is added to test that the hash is indeed only calculated once, and it contains a performance test, which will not even finish without caching.

We could also simplify the calculation of certain classes as well, e.g. ParseState, but I left that out for now. See [my comment here](https://github.com/parsingdata/metal/pull/403/files#r1386546508) for the performance comparison.

The build is :x:, because Sonar is not content  with this change, see [my comment here](https://github.com/parsingdata/metal/pull/403#issuecomment-1801855645). I would like to hear your thought about this as well.

Update: marked the Sonar remarks about unimplemented hashCode as false positives, since they are implemented, but Sonar is unable to detect it.
Update2: removed unused imports, and now we have 100% coverage again.

### CODE REVIEW

File: core/src/main/java/io/parsingdata/metal/data/ParseState.java [hunk 1]
```diff
@@ -116,7 +117,7 @@ public boolean equals(final Object obj) {
     }
 
     @Override
-    public int hashCode() {
+    public int cachingHashCode() {
         return Objects.hash(getClass(), order, offset, source, iterations, references);
```
Comment (mvanaken) [inline core/src/main/java/io/parsingdata/metal/data/ParseState.java:121]: We may consider doing something here like:
```suggestion
        return Objects.hash(getClass(), order.size, offset, source, iterations.size, references.size);
```
to keep hash calculations simple and quick.

I ran the performance test within ImmutableObjectTest for a `dataBlockCount` of 512 and switched between hash caching on/off and parsestate hash simplified on/off. That resulted into the following performance matrix:
|      | No hash caching  | Hash caching |
| --- | ------------- | ------------- |
| **Original ParseState hash** | undetermined | 730ms |
| **Simplified ParseState hash** | 3600ms | 700ms |

Since the hash caching contributes most to the performance gain, I implemented that first.

**Comment by mvanaken** [comment]
> So, how do we shut Sonar up? ;)

Discussed with @jvdb: we should mark these as false positives, because Sonar is unable to detect that the hashcode is also implemented. The ImmutableObject makes it mandatory.

File: core/src/main/java/io/parsingdata/metal/ImmutableObject.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * It will improve performance if these objects caches their hash.
+ * <p>
+ * This is a lazy implementation, instead of calculating the hash once within the constructor, to avoid
+ * performance decrease during parsing. In most implementations the hash is not actually used/needed.
+ */
+public abstract class ImmutableObject {
+
+    private Integer hash;
+
+    public abstract int cachingHashCode();
```
Comment (jvdb) [inline core/src/main/java/io/parsingdata/metal/ImmutableObject.java:14]: Maybe this should have a name that more reflects how an implementer should think about it? Such as `fixedHashCode()` or 'initFixedHashCode()` or `permanentHashCode()`?

Maybe `cachingHashCode()` explains that well, I'm just not sure about it because it's also about what it's used for (which is a kind of documentation, but also orthogonal to its functioning?)

File: core/src/main/java/io/parsingdata/metal/data/ParseState.java (hunk 1 above)
Comment (jvdb) [reply core/src/main/java/io/parsingdata/metal/data/ParseState.java:121]: Since the impact of hash caching itself is so huge, I would wait with further optimization until there is a clear practical case where that needs to be added. That said, I don't oppose this change if you want to add it.

File: core/src/test/java/io/parsingdata/metal/ImmutableObjectTest.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+import io.parsingdata.metal.data.ImmutableList;
+import io.parsingdata.metal.data.ParseState;
+import io.parsingdata.metal.data.ParseValue;
+import io.parsingdata.metal.data.Selection;
+import io.parsingdata.metal.expression.value.Value;
+import io.parsingdata.metal.token.Token;
+import io.parsingdata.metal.util.InMemoryByteStream;
+
+class ImmutableObjectTest {
```
Comment (jvdb) [inline core/src/test/java/io/parsingdata/metal/ImmutableObjectTest.java:35]: Nice tests!

**Review by jvdb — APPROVED** [review]
Solid improvement!

File: core/src/main/java/io/parsingdata/metal/ImmutableObject.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package io.parsingdata.metal;
+
+/**
+ * When objects are immutable, that means their hashcode will stay the same the moment it is created.
+ * It will improve performance if these objects caches their hash.
```
Comment (Carly-1) [inline core/src/main/java/io/parsingdata/metal/ImmutableObject.java:5]: Nitpick:
```suggestion
 * It will improve performance if these objects cache their hash.
```

File: core/src/main/java/io/parsingdata/metal/ImmutableObject.java (hunk 2 above)
Comment (Carly-1) [reply core/src/main/java/io/parsingdata/metal/ImmutableObject.java:14]: You could also add some javadoc to explain this more clearly? Not sure, maybe naming is enough.

File: core/src/test/java/io/parsingdata/metal/ImmutableObjectTest.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            @Override
+            public int cachingHashCode() {
+                counter.incrementAndGet();
+                return 15;
+            }
+        };
+        assertEquals(15, immutableObject.hashCode());
+        assertEquals(15, immutableObject.hashCode());
+        assertEquals(15, immutableObject.hashCode());
+        assertEquals(1, counter.get());
```
Comment (Carly-1) [inline core/src/test/java/io/parsingdata/metal/ImmutableObjectTest.java:67]: Nice 😄

**Review by Carly-1 — APPROVED** [review]
Very nice performance improvement!

File: core/src/main/java/io/parsingdata/metal/ImmutableObject.java (hunk 2 above)
Comment (mvanaken) [reply core/src/main/java/io/parsingdata/metal/ImmutableObject.java:14]: What do you think about `immutableHashCode` as alternative?

> jvdb: Ja, goed, goed!

Alrighty then!

File: core/src/main/java/io/parsingdata/metal/ImmutableObject.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
+package io.parsingdata.metal;
+
+/**
+ * When objects are immutable, that means their hashcode will stay the same the moment it is created.
```
Comment (rdvdijk) [inline core/src/main/java/io/parsingdata/metal/ImmutableObject.java:4]:
```suggestion
 * When objects are immutable, that means their hash code will stay the same the moment it is created.
```

File: core/src/main/java/io/parsingdata/metal/ImmutableObject.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+ * It will improve performance if these objects cache their hash.
```
Comment (rdvdijk) [inline core/src/main/java/io/parsingdata/metal/ImmutableObject.java:5]:
```suggestion
 * It will improve performance if these objects cache their hash code.
```

File: core/src/main/java/io/parsingdata/metal/ImmutableObject.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+ * <p>
+ * This is a lazy implementation, instead of calculating the hash once within the constructor, to avoid
+ * performance decrease during parsing. In most implementations the hash is not actually used/needed.
```
Comment (rdvdijk) [inline core/src/main/java/io/parsingdata/metal/ImmutableObject.java:8]:
```suggestion
 * performance decrease during parsing. In most implementations the hash code is not actually used/needed.
```

File: core/src/test/java/io/parsingdata/metal/data/SelectionTest.java [hunk 9]
```diff
@@ -41,6 +41,9 @@ private Source createSource() {
         return new Source() {
             @Override protected byte[] getData(BigInteger offset, BigInteger length) { return new byte[0]; }
             @Override protected boolean isAvailable(BigInteger offset, BigInteger length) { return true; }
+            @Override public int immutableHashCode() { return 0; }
+            @Override
+            public boolean equals(Object obj) {return obj == this;}
```
Comment (rdvdijk) [inline core/src/test/java/io/parsingdata/metal/data/SelectionTest.java:46]: Minor style consistency nitpick:

```suggestion
            public boolean equals(Object obj) { return obj == this; }
```

File: core/src/test/java/io/parsingdata/metal/data/selection/ByTypeTest.java [hunk 10]
```diff
@@ -38,6 +38,9 @@ public class ByTypeTest {
     public static final Source EMPTY_SOURCE = new Source() {
         @Override protected byte[] getData(BigInteger offset, BigInteger length) { throw new IllegalStateException(); }
         @Override protected boolean isAvailable(BigInteger offset, BigInteger length) { return false; }
+        @Override public int immutableHashCode() { return -1; }
+        @Override
+        public boolean equals(Object obj) {return obj == this;}
```
Comment (rdvdijk) [inline core/src/test/java/io/parsingdata/metal/data/selection/ByTypeTest.java:43]: See [this suggestion](https://github.com/parsingdata/metal/pull/403/files#r1404326638).

File: core/src/test/java/io/parsingdata/metal/ImmutableObjectTest.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    @Test
+    void performanceTest() {
+        // This test would take way too much time without hash caching.
```
[the remark this answers is not in the document]
Comment (rdvdijk) [reply core/src/test/java/io/parsingdata/metal/ImmutableObjectTest.java:72]: Ah, `@Timeout`: https://www.baeldung.com/java-junit-5-timeout-annotation

Maybe add it, to make sure we never have a regression here?

File: core/src/main/java/io/parsingdata/metal/ImmutableObject.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+ */
+public abstract class ImmutableObject {
+
+    private Integer hash;
```
[the remark this answers is not in the document]
Comment (mvanaken) [reply core/src/main/java/io/parsingdata/metal/ImmutableObject.java:12]: Sure! I'll change it!

**Review by rdvdijk — APPROVED** [review]
Added some comments for minor improvements.

File: core/src/main/java/io/parsingdata/metal/ImmutableObject.java (hunk 12 above)
Comment (rdvdijk) [inline core/src/main/java/io/parsingdata/metal/ImmutableObject.java:12]: Another consideration: This is an `Integer`, and not an `int`, because it can be `null`.

Using boxed primitives isn't "free", though. The `int` would be 4 bytes, but each boxed `Integer` is 16 or even 24 bytes (according to [this Stackoverflow answer](https://stackoverflow.com/a/9037808)).

One alternative here could be to have a primitive `int hashCode` and a `boolean initialized`.

Not sure if this is worth the effort, but since most classes now inherit from this one, and they are used all over the `ParseGraph`, it might be worth considering.

This would be an easy fix in the future, so for now I think it's okay to keep this current implementation.

File: core/src/test/java/io/parsingdata/metal/ImmutableObjectTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (mvanaken) [reply core/src/test/java/io/parsingdata/metal/ImmutableObjectTest.java:72]: I did add that, but I have removed it again for two reasons:
1. The timing specified is an absolute timing and this test is dependent on the agent on GitHub, which could fluctuate. So that number would be arbitrary.
2. The TimeOut annotation does not stop the test. It simply checks if the time it took exceeds the timeout value. That means that the run would still run a very long time in case the hash cache is turned off and that would most likely trigger a GitHub error sooner than it would trigger the timeout error.

[the remark this answers is not in the document]
Comment (mvanaken) [reply core/src/test/java/io/parsingdata/metal/ImmutableObjectTest.java:72]: Discussed it again with @jvdb. I'm going to tune the numbers in this test, such that without hash cashing it would take e..g 30 seconds - 1 minute to run, while with hash cashing in a few milliseconds. I could then add a TimeOut annotation with a limit of e.g. 10 seconds. Then we would not have an error build when there is a GitHub glitch or something, and the test would also not run for a very long time in case hash cashing is (accidentally) turned off.

File: core/src/main/java/io/parsingdata/metal/ImmutableObject.java (hunk 12 above)
Comment (mvanaken) [reply core/src/main/java/io/parsingdata/metal/ImmutableObject.java]: Ah, interesting point! I found the use of an object here quite elegant, opposed to a separate boolean, but I haven't considered how much memory this would need. There are two concerns here:
1. Performance
2. Memory usage

1. Performance: I have run the performance test for large numbers and there does not seem to be a difference when using int over Integer. I think the JVM does a lot of auto improvements here.
2. Memory usage: hashCodes are not calculated for all objects. They are only calculated when needed. Since they are initialized as null, they only require as much space as a pointer requires, which is 4 bytes like a primitive int. When we would use ints + booleans as fields, then this requires a little more memory to store the initialize boolean variable. Assuming a boolean requires a bit of space (i mean this literal ;)), and a Integer hashCode requires 160 bits additional bits when calculated, then that means that if we calculate less then 1/160th of the total number of ImmutableObjects, then the Integer implementation actually requires less memory.


Tl;dr;
So, in short, performance is not an issue and since memory usage depends on the implementation on how many objects you want to calculates hashes for (and the default is for 0 objects), it may be wise to keep this as is.

**Comment by mvanaken** [comment]
I'm going to merge this branch even though the codecov analysis failed. That analysis is incorrect, claiming that we have less coverage, but we do not. The analysis on the main branch will be green. If not, pie for everyone, my treat!
